Implement named inference rule for torch.sum

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23316 Add names to repr for named tensors
* #23315 Rename tensor.is_named to has_named, expose has_named to python.
* #23237 Enable named tensors for arithmetic, clone, and tensor conversion ops
* #23229 Named inference rule for copy_
* #23193 Implement named inference rule for mul
* #23172 Implement tensor.set_names_, tensor.names setter
* #23106 Named inference rule for torch.prod
* **#23081 Implement named inference rule for torch.sum**

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 23081 gh/zou3519/71/head

Differential Revision: [D16419174](https://our.internmc.facebook.com/intern/diff/D16419174)

zou3519 committed:
Implement named inference rule for torch.sum

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

zou3519 committed:
Update on "Implement named inference rule for torch.sum"

Implement named inference rule for torch.sum

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 23081 gh/zou3519/71/head

zou3519 committed:
Update on "Implement named inference rule for torch.sum"

Implement named inference rule for torch.sum

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 23081 gh/zou3519/71/head

zou3519 committed:
Update on "Implement named inference rule for torch.sum"

Implement named inference rule for torch.sum

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 23081 gh/zou3519/71/head

nairbv commented on aten/src/ATen/NamedTensorUtils.h:
```diff
@@ -37,6 +38,7 @@ void propagate_names(Tensor& result, const Tensor& src);
 void propagate_names(TensorImpl* result, /*const */TensorImpl* src);
 
 void propagate_names_except(Tensor& result, const Tensor& src, IntArrayRef excluded_idxs);
+void propagate_names_for_reduction(Tensor& result, const Tensor& src, IntArrayRef excluded_idxs, bool keepdim);
```
might be helpful to have a code comment explaining the different name propagation semantics and when/where each of these three are used

zou3519 committed:
Update on "Implement named inference rule for torch.sum"

Implement named inference rule for torch.sum

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 23081 gh/zou3519/71/head

zou3519 committed:
Update on "Implement named inference rule for torch.sum"

Implement named inference rule for torch.sum

Featuring:
- Changes to PythonArgParser and the codegen to support DimnameList[1].
This is a union type: Union[Dimname, List[Dimname]]

Test Plan:
- New tests [namedtensor ci]

gh-metadata: pytorch pytorch 23081 gh/zou3519/71/head
